## Homework #3 

2. Reconsider the pig problem of Example 1.1, but now suppose that the
weight of the pig after t days is w = 800=(1 + 3e−t=30) lbs.

**Example 1.1**: "A pig weighing 200 pounds gains 5 pounds per day and costs
45 cents a day to keep. The market price for pigs is 65 cents per pound, but is
falling 1 cent per day. When should the pig be sold?"


 b) Find the optimal time to sell the pig. Use the five-step approach,
and model as a one–variable optimization problem.


In [3]:
#Newton's method to solve (b)
#f is the function we are evalating, df is the derivative of that function
#steps is number of iterations, x0 is the starting location
function Newtons(f,df, steps, x0)
    x = x0
    
    for i in 1:steps
        
        x = x - f(x)/df(x)
    
    end
    println(x)
    return x
end

Newtons (generic function with 1 method)

Newton's method is used on the first derivative of the profit function to find the best time to sell.Profit function has one variable, x, the day the pig is sold. It returns the profit of the sale. Below $F(x)$ is the first derivative of the Profit function and $DF(x)$ is the second derivative.

In [4]:
# first derivative of our profit function
F(x) = (140*e^(-x/30) - 4*e^(-x/30)*x -40)/(5*(1+3*e^(-x/30))^2) -0.45

# second derivative of the profit fuction
DF(x) = (90*e^(-x/10) +2*e^(-x/30)*x - 250*e^(-x/30) - 720*e^(-x/15) - 18e^(-x/10)*x)/(75*(3*e^(-x/30)+1)^4)

#Use 12 steps which is more than enough
Newtons(F,DF, 12,14)

12.334891442632157


12.334891442632157

We can't sell at a fraction of a day so we round down to 12. The optimal time to sell the pig is day 12.

**(c)**
The parameter 800 represents the eventual mature weight of the pig.
Perform a sensitivity analysis for this parameter. Consider both the
best time to sell and the profit obtained.

we need to write the function with var i in place of 800.

Also redo Newton's so the functions can have two inputs

In [10]:
g(t,i) = ((0.65 - 0.01*t)*i)/(1+3*e^(-t/30)) - 0.45 *t

Dg(t,i) = ((0.035*e^(-t/30) - 0.001*e^(-t/30)*t - 0.01)*i)/((1+3*e^(-t/30))^2) - 0.45

DDg(t,i) = ((0.0015*e^(-t/10) + 0.00003*e^(-t/30)*t - 0.012*e^(-t/15) - 0.00417*e^(-t/30) - 0.0003*e^(-t/10)*t)*i)/((1+3*e^(-t/30))^4)



DDg (generic function with 1 method)

Newton's method with 2 variables, this allows us to change the value of i (mature weight of pig) while solving for optimal time 
to sell. So only the value t changes with each iteration

In [7]:
#t0 starting point, i-mature weight of pig
function Newtons_2var(f,df,t0, steps, i)
    t = t0
    u = 0
    
    for n = 1:steps
        u = t
        t = t- f(t,i)/df(t,i)
        
        if u == t
            return t
        end
        
    end
    
    return t
end

Newtons_2var (generic function with 1 method)

The sensitivity is given by the equation

$S(t,i) = dt/di * i/t$

This function calculates S(t,i) by approximating the derivative by evaluating the optimal time to sell for two different values of i (mature pig weight), these are given below by t and tt, then finding the change in rate of the optimal time to sell.

In [20]:
#t0- starting point for the two newton's methods, steps- number of steps taken 
#cc- the difference between the two calculatoins
function sensitivity_wrt_t(t0,steps, cc)
    #t is the optimal time to sell when i = 800
    t = Newtons_2var(Dg,DDg,t0,steps, 800)
    
    #tt is the optimal time to seel when i = 800 + cc (some small amount)
    tt = Newtons_2var(Dg,DDg,t0,steps, 800 + cc)
    
    #returns the two optimal times (t, tt)
    print(t, "  ", tt)
    
    #Derivative (dt/di) given by (tt-t)/(cc)
    return (tt-t)/(cc) *800/t
end

sensitivity_wrt_t (generic function with 1 method)

In [19]:
sensitivity_wrt_t(12,25,0.00001)

12.334891442632161  12.334891509385942

0.4329428044330625

The sensitivity of the optimal time to sell is 0.432, this implies that if you increased the mature pig weight (i) by 10% the optimal time to sell would increase by 4.3%

Calculates the sensitivity of the Profits based on the mature pig weight in the same fashion

In [32]:
#Calculates S(P,i)
function sensitivity_wrt_P(t0,steps, cc)
    #t is the optimal time to sell when i = 800
    t = Newtons_2var(Dg,DDg,t0,steps, 800)
    
    #tt is the optimal time to seel when i = 800 + cc (some small amount)
    tt = Newtons_2var(Dg,DDg,t0,steps, 800 + cc)
    
    print(t, "  ", tt)
    
    #g(tt,800+cc) calculates the optimal profit when mature pig weight is 800 + cc (some small amount)
    #g(t,800) calculates the optimal profit when mature pig weight is 800 
    return (g(tt,800+cc)-g(t,800))/(cc) *800/g(t,800)
end

sensitivity_wrt_P (generic function with 1 method)

In [35]:
sensitivity_wrt_P(12,25, 0.0001)

12.334891442632161  12.334892110169902

1.040987704090318

The sensitivity of the profit based on mature pig weight is 1.04, this implies that if you increased the mature pig weight (i) by 10% the profit would increase by 10.4%